packages/permission-controller/src/PermissionController.ts

```diff
@@ -1976,21 +1996,63 @@ export class PermissionController<
    * @param permissionsRequest - The permissions request object.
    * @returns The approved permissions request object.
    */
-  private async requestUserApproval(permissionsRequest: PermissionsRequest) {
+  private async requestUserApproval(
+    permissionsRequest: PermissionsRequest,
+  ): Promise<{
+    approvedRequest: PermissionsRequest;
+    sideEffectsData?: unknown[];
+  }> {
     const { origin, id } = permissionsRequest.metadata;
+
+    const sideEffects = Object.keys(
+      permissionsRequest.permissions,
+    ).reduce<SideEffects>(
+      (sideEffectList, targetName) => {
+        const targetKey = this.getTargetKey(targetName);
+
+        if (targetKey) {
+          const specification = this.getPermissionSpecification(targetKey);
+
+          if (specification.sideEffect) {
+            sideEffectList.permittedHandlers.push(
+              specification.sideEffect.onPermitted,
+            );
+            sideEffectList.failureHandlers.push(
+              specification.sideEffect.onFailure,
+            );
+          }
+        }
+        return sideEffectList;
+      },
+      { permittedHandlers: [], failureHandlers: [] },
+    );
+
+    const hasSideEffects = sideEffects.permittedHandlers.length !== 0;
+
     const approvedRequest = await this.messagingSystem.call(
       'ApprovalController:addRequest',
       {
         id,
         origin,
         requestData: permissionsRequest,
         type: MethodNames.requestPermissions,
+        sideEffects: hasSideEffects ? sideEffects : undefined,
```

Comment:
Forgive me if I'm missing the obvious, but could this not just be implemented directly in the `PermissionController` by just invoking the `onPermitted` handlers after a successful `messagingSystem.call` and then calling the `onFailure` handlers for any of those callbacks that fail?

I'm not sure what we're gaining by adding the side effect support to the `ApprovalController` as it already returns a promise so does it not already support the callbacks we or anyone else would need via `await` or `then` and `catch`?

My worry would be we're complicating its API as the user would have to choose between Promises or callbacks to control the flow.

Reply:
@Gudahtt and I discussed it a long time ago, IIRC it's to also support side-effects for non-permissions approvals (like adding a new chain, that was the example @Gudahtt gave me I think). It also implements the support in the PermissionController because we need it at this level for Snaps

Reply:
> But my main concern is I fear we're adding complexity to the API of the ApprovalController that it doesn't require to support the use cases we may want in future

I'll let @Gudahtt answer on this one

> where we could use native features like Promise.all with multiple approval requests for example
I don't see where this would resolve the issue side-effects need to resolve

Reply:
If the confirmations systems team would prefer just handling the side-effects in the PermissionController for now I would say that is fine. That still solves our problems and seems cleaner. 

Side-effects to approvals may be something that should be handled differently? Regardless, they will own adding that when it is needed, so 🤷 Would be nice with a unified system though..

Reply:
As per today's meeting we will handle side-effects in the PermissionController for now.